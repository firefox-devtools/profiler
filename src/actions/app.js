/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { oneLine } from 'common-tags';
import {
  getSelectedTab,
  getDataSource,
  getLocalTrackOrderByPid,
} from 'firefox-profiler/selectors/url-state';
import {
  getTrackThreadHeights,
  getIsEventDelayTracksEnabled,
  getIsExperimentalCPUGraphsEnabled,
  getIsExperimentalProcessCPUTracksEnabled,
} from 'firefox-profiler/selectors/app';
import {
  getLocalTracksByPid,
  getThreads,
  getCounters,
} from 'firefox-profiler/selectors/profile';
import { sendAnalytics } from 'firefox-profiler/utils/analytics';
import {
  stateFromLocation,
  withHistoryReplaceStateSync,
} from 'firefox-profiler/app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { fatalError } from './errors';
import {
  addEventDelayTracksForThreads,
  initializeLocalTrackOrderByPid,
  addProcessCPUTracksForProcess,
} from 'firefox-profiler/profile-logic/tracks';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getIsCPUUtilizationProvided,
  getAreThereAnyProcessCPUCounters,
} from 'firefox-profiler/selectors/cpu';

import type {
  Profile,
  ThreadsKey,
  CssPixels,
  Action,
  ThunkAction,
  UrlState,
  UploadedProfileInformation,
  IndexIntoCategoryList,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnection,
  BrowserConnectionStatus,
} from 'firefox-profiler/app-logic/browser-connection';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 * `profile` parameter can be null when the data source can't provide the profile
 * and the url upgrader step is not needed (e.g. 'from-browser').
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile | null,
  browserConnection: BrowserConnection | null
): ThunkAction<void> {
  return (dispatch) => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      if (e.name === 'UrlUpgradeError') {
        // The error is an URL upgrade error, let's fire a fatal error.
        // If there's a service worker update, the class `ServiceWorkerManager`
        // will automatically reload in case the new code knows how to handle
        // this URL version.
        dispatch(fatalError(e));
        return;
      }
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    // All of this is done while the history is replaced, as this is part of the initial
    // load process.
    withHistoryReplaceStateSync(() => {
      dispatch(updateUrlState(urlState));
      dispatch(finalizeProfileView(browserConnection));
      dispatch(urlSetupDone());
    });
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadsKey: ThreadsKey,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadsKey];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadsKey,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}

/*
 * This action enables the event delay tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableEventDelayTracks()`
 */
export function enableEventDelayTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsEventDelayTracksEnabled(getState())) {
      console.error(
        'Tried to enable the event delay tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      selectedThreadSelectors.getSamplesTable(getState()).eventDelay ===
      undefined
    ) {
      // Return early if the profile doesn't have eventDelay values.
      console.error(oneLine`
        Tried to enable the event delay tracks, but this profile does
        not have eventDelay values. It is likely an older profile.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addEventDelayTracksForThreads(
      getThreads(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );
    dispatch({
      type: 'ENABLE_EVENT_DELAY_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/*
 * This action enables the CPU graph tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableCPUGraphs()`
 */
export function enableExperimentalCPUGraphs(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalCPUGraphsEnabled(getState())) {
      console.error(
        'Tried to enable the CPU graph tracks, but they are already enabled.'
      );
      return false;
    }

    if (!getIsCPUUtilizationProvided(getState())) {
      // Return early if the profile doesn't have threadCPUDelta values.
      console.error(oneLine`
        Tried to enable the CPU graph tracks, but this profile does
        not have threadCPUDelta values. It is likely an older profile.
      `);
      return false;
    }

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_CPU_GRAPHS',
    });

    return true;
  };
}

/*
 * This action enables the process CPU tracks. They are hidden by default because
 * the front-end work is not done for this data yet.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableExperimentalProcessCPUTracks()`
 */
export function enableExperimentalProcessCPUTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalProcessCPUTracksEnabled(getState())) {
      console.error(
        'Tried to enable the process CPU tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      !getIsCPUUtilizationProvided(getState()) &&
      getAreThereAnyProcessCPUCounters(getState())
    ) {
      // Return early if the profile doesn't have threadCPUDelta values or
      // doesn't have any experimental process CPU counters.
      console.error(oneLine`
        Tried to enable the process CPU tracks, but this profile does
        not have threadCPUDelta values or process CPU threads.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addProcessCPUTracksForProcess(
      getCounters(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/**
 * This caches the profile data in the local state for synchronous access.
 */
export function setCurrentProfileUploadedInformation(
  uploadedProfileInformation: UploadedProfileInformation | null
): Action {
  return {
    type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
    uploadedProfileInformation,
  };
}

export function profileRemotelyDeleted(): Action {
  // Ideally we should store the current profile data in a local indexeddb, and
  // set the URL to /local/<indexeddb-key>.
  return { type: 'PROFILE_REMOTELY_DELETED' };
}

export function updateBrowserConnectionStatus(
  browserConnectionStatus: BrowserConnectionStatus
): Action {
  return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
}

export function toggleOpenCategoryInSidebar(
  kind: string,
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    kind,
    category,
  };
}
