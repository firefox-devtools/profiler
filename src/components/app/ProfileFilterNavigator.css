/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.profileFilterNavigator--tab-selector {
  display: flex;
  align-items: center;
  column-gap: 5px;

  /* Padding for the arrow on the left side and a bit on the other for hover. */
  padding-inline: 20px 5px;
}

/* This is the dropdown arrow on the left of the button. */
.profileFilterNavigator--tab-selector::before {
  position: absolute;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin: 5px 5px 0;
  color: var(--internal-selected-color);
  content: '';

  /* opt-out of forced color so the arrow is visible */
  forced-color-adjust: none;
  inset-block-start: 4px;
  inset-inline-start: 4px;
}

span.profileFilterNavigator--tab-selector::before {
  /* Disabled tab selector indicates this with a grayed out arrow. */
  color: var(--grey-30);
}

@media (forced-colors: active) {
  .profileFilterNavigator--tab-selector::before {
    color: currentcolor;
  }

  span.profileFilterNavigator--tab-selector::before {
    color: GrayText;
  }
}
