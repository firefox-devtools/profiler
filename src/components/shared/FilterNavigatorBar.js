/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import './FilterNavigatorBar.css';

type FilterNavigatorBarListItemProps = {
  +onClick?: null | ((number) => mixed),
  +index: number,
  +isFirstItem: boolean,
  +isLastItem: boolean,
  +isSelectedItem: boolean,
  +title?: string,
  +additionalClassName?: string,
  +children: React.Node,
};

class FilterNavigatorBarListItem extends React.PureComponent<FilterNavigatorBarListItemProps> {
  _onClick = () => {
    const { index, onClick } = this.props;
    if (onClick) {
      onClick(index);
    }
  };

  render() {
    const {
      isFirstItem,
      isLastItem,
      isSelectedItem,
      children,
      additionalClassName,
      onClick,
      title,
    } = this.props;
    return (
      <li
        className={classNames('filterNavigatorBarItem', additionalClassName, {
          filterNavigatorBarRootItem: isFirstItem,
          filterNavigatorBarSelectedItem: isSelectedItem,
          filterNavigatorBarLeafItem: isLastItem,
        })}
        title={title}
        onClick={onClick ? this._onClick : null}
      >
        {onClick ? (
          <button type="button" className="filterNavigatorBarItemContent">
            {children}
          </button>
        ) : (
          <span className="filterNavigatorBarItemContent">{children}</span>
        )}
      </li>
    );
  }
}

type Props = {
  +className: string,
  +items: $ReadOnlyArray<React.Node>,
  +onPop: (number) => mixed,
  +selectedItem: number,
  +uncommittedItem?: string,
};

export class FilterNavigatorBar extends React.PureComponent<Props> {
  render() {
    const { className, items, selectedItem, uncommittedItem, onPop } =
      this.props;
    return (
      <TransitionGroup
        component="ol"
        className={classNames('filterNavigatorBar', className)}
      >
        {items.map((item, i) => (
          <CSSTransition
            key={i}
            classNames="filterNavigatorBarTransition"
            timeout={250}
          >
            <FilterNavigatorBarListItem
              index={i}
              onClick={
                i === items.length - 1 && !uncommittedItem ? null : onPop
              }
              isFirstItem={i === 0}
              isLastItem={i === items.length - 1}
              isSelectedItem={i === selectedItem}
            >
              {item}
            </FilterNavigatorBarListItem>
          </CSSTransition>
        ))}
        {uncommittedItem ? (
          <CSSTransition
            key={items.length}
            classNames="filterNavigatorBarUncommittedTransition"
            timeout={0}
          >
            <FilterNavigatorBarListItem
              index={items.length}
              isFirstItem={false}
              isLastItem={true}
              isSelectedItem={false}
              additionalClassName="filterNavigatorBarUncommittedItem"
              title={uncommittedItem}
            >
              {uncommittedItem}
            </FilterNavigatorBarListItem>
          </CSSTransition>
        ) : null}
      </TransitionGroup>
    );
  }
}
