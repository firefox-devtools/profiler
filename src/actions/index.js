/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
