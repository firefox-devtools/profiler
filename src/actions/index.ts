/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
