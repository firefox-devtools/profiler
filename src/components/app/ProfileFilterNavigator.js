/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import memoize from 'memoize-immutable';
import { Localized } from '@fluent/react';
import { showMenu } from '@firefox-devtools/react-contextmenu';
import classNames from 'classnames';

import explicitConnect from 'firefox-profiler/utils/connect';
import { popCommittedRanges } from 'firefox-profiler/actions/profile-view';
import {
  getPreviewSelection,
  getProfileFilterPageData,
  getProfileFilterPageDataByTabID,
  getProfileRootRange,
  getProfileTimelineUnit,
  getCommittedRangeLabels,
} from 'firefox-profiler/selectors/profile';
import { getTabFilter } from 'firefox-profiler/selectors/url-state';
import { getFormattedTimelineValue } from 'firefox-profiler/profile-logic/committed-ranges';
import { FilterNavigatorBar } from 'firefox-profiler/components/shared/FilterNavigatorBar';
import { Icon } from 'firefox-profiler/components/shared/Icon';
import { TabSelectorMenu } from '../shared/TabSelectorMenu';

import type { ElementProps } from 'react';
import type {
  ProfileFilterPageData,
  StartEndRange,
  TabID,
} from 'firefox-profiler/types';

import './ProfileFilterNavigator.css';

type Props = {|
  +filterPageDataForActiveTab: ProfileFilterPageData | null,
  +pageDataByTabID: Map<TabID, ProfileFilterPageData> | null,
  +tabFilter: TabID | null,
  +rootRange: StartEndRange,
  +profileTimelineUnit: string,
  ...ElementProps<typeof FilterNavigatorBar>,
|};

type DispatchProps = {|
  +onPop: $PropertyType<Props, 'onPop'>,
|};

type StateProps = $ReadOnly<$Exact<$Diff<Props, DispatchProps>>>;

class ProfileFilterNavigatorBarImpl extends React.PureComponent<Props> {
  _getItemsWithFirstElement = memoize(
    (firstItem, items) => [firstItem, ...items],
    {
      limit: 1,
    }
  );

  _showTabSelectorMenu = (event: SyntheticMouseEvent<HTMLElement>) => {
    if (this.props.items.length > 0 || this.props.uncommittedItem) {
      // Do nothing if there are committed ranges. We only allow users to change
      // the tab if they are on root range.
      return;
    }

    const rect = event.currentTarget.getBoundingClientRect();
    showMenu({
      data: null,
      id: 'TabSelectorMenu',
      position: { x: rect.left, y: rect.bottom },
      target: event.target,
    });
  };

  render() {
    const {
      className,
      items,
      selectedItem,
      uncommittedItem,
      onPop,
      rootRange,
      filterPageDataForActiveTab,
      pageDataByTabID,
      tabFilter,
      profileTimelineUnit,
    } = this.props;

    let firstItem;
    if (filterPageDataForActiveTab) {
      // TODO: Remove this once we ship the tab selector and remove the active tab view.
      firstItem = (
        <>
          {filterPageDataForActiveTab.favicon ? (
            <Icon iconUrl={filterPageDataForActiveTab.favicon} />
          ) : null}
          <span title={filterPageDataForActiveTab.origin}>
            {filterPageDataForActiveTab.hostname} (
            {getFormattedTimelineValue(
              rootRange.end - rootRange.start,
              profileTimelineUnit
            )}
            )
          </span>
        </>
      );
    } else {
      // pageDataByTabID will be empty if there is no page information in the
      // profile or when the page information is empty. This could happen for
      // older profiles and profiles from external importers that don't have
      // this information.
      if (pageDataByTabID && pageDataByTabID.size > 0) {
        const pageData =
          tabFilter !== null ? pageDataByTabID.get(tabFilter) : null;

        const itemContents = pageData ? (
          <>
            {/* Show the page data if the profile is filtered by tab */}
            {pageData.favicon ? <Icon iconUrl={pageData.favicon} /> : null}
            <span title={pageData.origin}>
              {pageData.hostname} (
              {getFormattedTimelineValue(
                rootRange.end - rootRange.start,
                profileTimelineUnit
              )}
              )
            </span>
          </>
        ) : (
          <Localized
            id="ProfileFilterNavigator--full-range-with-duration"
            vars={{
              fullRangeDuration: getFormattedTimelineValue(
                rootRange.end - rootRange.start,
                profileTimelineUnit
              ),
            }}
          >
            Full Range
          </Localized>
        );

        if (items.length === 0 && !uncommittedItem) {
          // It should be a clickable button if there are no committed ranges.
          firstItem = (
            <button
              type="button"
              onClick={this._showTabSelectorMenu}
              className={classNames(
                'filterNavigatorBarItemContent',
                'profileFilterNavigator--tab-selector'
              )}
            >
              {itemContents}
            </button>
          );
        } else {
          // There are committed ranges, don't make it button because this will
          // be wrapped with a button.
          firstItem = (
            <span
              className={classNames(
                'filterNavigatorBarItemContent',
                'profileFilterNavigator--tab-selector'
              )}
            >
              {itemContents}
            </span>
          );
        }
      } else {
        firstItem = (
          <Localized
            id="ProfileFilterNavigator--full-range-with-duration"
            vars={{
              fullRangeDuration: getFormattedTimelineValue(
                rootRange.end - rootRange.start,
                profileTimelineUnit
              ),
            }}
          >
            Full Range
          </Localized>
        );
      }
    }

    const itemsWithFirstElement = this._getItemsWithFirstElement(
      firstItem,
      items
    );
    return (
      <>
        <FilterNavigatorBar
          className={className}
          items={itemsWithFirstElement}
          selectedItem={selectedItem}
          uncommittedItem={uncommittedItem}
          onPop={onPop}
        />
        {pageDataByTabID && pageDataByTabID.size > 0 ? (
          <TabSelectorMenu />
        ) : null}
      </>
    );
  }
}

export const ProfileFilterNavigator = explicitConnect<
  {||},
  StateProps,
  DispatchProps,
>({
  mapStateToProps: (state) => {
    const items = getCommittedRangeLabels(state);
    const previewSelection = getPreviewSelection(state);
    const profileTimelineUnit = getProfileTimelineUnit(state);
    const uncommittedItem = previewSelection.hasSelection
      ? getFormattedTimelineValue(
          previewSelection.selectionEnd - previewSelection.selectionStart,
          profileTimelineUnit
        )
      : undefined;

    // TODO: Remove this once we ship the tab selector and remove the active tab view.
    const filterPageDataForActiveTab = getProfileFilterPageData(state);
    const pageDataByTabID = getProfileFilterPageDataByTabID(state);
    const tabFilter = getTabFilter(state);
    const rootRange = getProfileRootRange(state);
    return {
      className: 'profileFilterNavigator',
      items: items,
      // Do not remove 1 from the length because we are going to increment this
      // array's length by adding the first element.
      selectedItem: items.length,
      uncommittedItem,
      filterPageDataForActiveTab,
      pageDataByTabID,
      tabFilter,
      rootRange,
      profileTimelineUnit,
    };
  },
  mapDispatchToProps: {
    onPop: popCommittedRanges,
  },
  component: ProfileFilterNavigatorBarImpl,
});
