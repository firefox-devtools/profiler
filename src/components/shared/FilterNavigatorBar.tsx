/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import * as React from 'react';
import classNames from 'classnames';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import './FilterNavigatorBar.css';

type FilterNavigatorBarListItemProps = {
  readonly onClick?: null | ((index: number) => unknown);
  readonly index: number;
  readonly isFirstItem: boolean;
  readonly isLastItem: boolean;
  readonly isSelectedItem: boolean;
  readonly title?: string;
  readonly additionalClassName?: string;
  readonly children: React.ReactNode;
};

class FilterNavigatorBarListItem extends React.PureComponent<FilterNavigatorBarListItemProps> {
  _onClick = () => {
    const { index, onClick } = this.props;
    if (onClick) {
      onClick(index);
    }
  };

  override render() {
    const {
      isFirstItem,
      isLastItem,
      isSelectedItem,
      children,
      additionalClassName,
      onClick,
      title,
    } = this.props;
    return (
      <li
        className={classNames('filterNavigatorBarItem', additionalClassName, {
          filterNavigatorBarRootItem: isFirstItem,
          filterNavigatorBarSelectedItem: isSelectedItem,
          filterNavigatorBarLeafItem: isLastItem,
        })}
        title={title}
        onClick={onClick ? this._onClick : null}
      >
        {onClick ? (
          <button type="button" className="filterNavigatorBarItemContent">
            {children}
          </button>
        ) : (
          <span className="filterNavigatorBarItemContent">{children}</span>
        )}
      </li>
    );
  }
}

type Props = {
  readonly className: string;
  readonly items: ReadonlyArray<React.ReactNode>;
  readonly onPop: (param: number) => mixed;
  readonly selectedItem: number;
  readonly uncommittedItem?: string;
};

export class FilterNavigatorBar extends React.PureComponent<Props> {
  override render() {
    const { className, items, selectedItem, uncommittedItem, onPop } =
      this.props;

    const transitions = items.map((item, i) => (
      <CSSTransition
        key={i}
        classNames="filterNavigatorBarTransition"
        timeout={250}
      >
        <FilterNavigatorBarListItem
          index={i}
          onClick={i === items.length - 1 && !uncommittedItem ? null : onPop}
          isFirstItem={i === 0}
          isLastItem={i === items.length - 1}
          isSelectedItem={i === selectedItem}
        >
          {item}
        </FilterNavigatorBarListItem>
      </CSSTransition>
    ));

    if (uncommittedItem) {
      transitions.push(
        <CSSTransition
          key={items.length}
          classNames="filterNavigatorBarUncommittedTransition"
          timeout={0}
        >
          <FilterNavigatorBarListItem
            index={items.length}
            isFirstItem={false}
            isLastItem={true}
            isSelectedItem={false}
            additionalClassName="filterNavigatorBarUncommittedItem"
            title={uncommittedItem}
          >
            {uncommittedItem}
          </FilterNavigatorBarListItem>
        </CSSTransition>
      );
    }

    return (
      <TransitionGroup
        component="ol"
        className={classNames('filterNavigatorBar', className)}
      >
        {transitions}
      </TransitionGroup>
    );
  }
}
