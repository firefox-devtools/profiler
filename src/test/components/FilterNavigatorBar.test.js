/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { Provider } from 'react-redux';

import {
  render,
  fireEvent,
  screen,
  act,
} from 'firefox-profiler/test/fixtures/testing-library';
import { FilterNavigatorBar } from 'firefox-profiler/components/shared/FilterNavigatorBar';
import { ProfileFilterNavigator } from '../../components/app/ProfileFilterNavigator';
import * as ProfileView from '../../actions/profile-view';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';
import { ensureExists } from '../../utils/flow';

describe('shared/FilterNavigatorBar', () => {
  it(`pops the item unless the last one is clicked`, () => {
    const onPop = jest.fn();
    render(
      <FilterNavigatorBar
        className=""
        items={['foo', 'bar']}
        selectedItem={2}
        onPop={onPop}
      />
    );

    // We don't use getByRole because this isn't a button.
    const lastElement = screen.getByText('bar');
    fireEvent.click(lastElement);
    expect(onPop).not.toHaveBeenCalled();

    const firstElement = screen.getByRole('button', { name: 'foo' });
    fireEvent.click(firstElement);
    expect(onPop).toHaveBeenCalledWith(0);
  });

  it(`pops the last item if there's an uncommited item`, () => {
    const onPop = jest.fn();
    render(
      <FilterNavigatorBar
        className=""
        items={['foo', 'bar']}
        selectedItem={2}
        uncommittedItem="baz"
        onPop={onPop}
      />
    );

    // We don't use getByRole because this isn't a button.
    const lastElement = screen.getByText('bar');
    fireEvent.click(lastElement);
    expect(onPop).toHaveBeenCalledWith(1);
  });
});

describe('app/ProfileFilterNavigator', () => {
  const tabID = 123123;
  function setup() {
    const { profile } = getProfileFromTextSamples(`
      A  A  A
      B  B  B
      C  C  H
      D  F  I
      E  G
      `);
    profile.pages = [
      {
        tabID: tabID,
        innerWindowID: 1,
        url: 'https://developer.mozilla.org/en-US/',
        embedderInnerWindowID: 0,
      },
    ];
    profile.meta.configuration = {
      threads: [],
      features: [],
      capacity: 1000000,
      activeTabID: tabID,
    };

    // Change the root range for testing.
    const samples = profile.threads[0].samples;
    ensureExists(samples.time)[samples.length - 1] = 50;

    const store = storeWithProfile(profile);
    const renderResult = render(
      <Provider store={store}>
        <ProfileFilterNavigator />
      </Provider>
    );

    return {
      ...store,
      ...renderResult,
    };
  }

  it('renders ProfileFilterNavigator properly', () => {
    const { container, dispatch } = setup();
    // Just root range
    expect(container.firstChild).toMatchSnapshot();

    // With committed range
    act(() => {
      dispatch(ProfileView.commitRange(0, 40));
    });
    expect(container.firstChild).toMatchSnapshot();

    // With preview selection
    act(() => {
      dispatch(
        ProfileView.updatePreviewSelection({
          hasSelection: true,
          isModifying: false,
          selectionStart: 10,
          selectionEnd: 10.1,
        })
      );
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the "Full Range" text as its first element', () => {
    const { getByText } = setup();
    expect(getByText(/Full Range/)).toBeInTheDocument();
  });
});
