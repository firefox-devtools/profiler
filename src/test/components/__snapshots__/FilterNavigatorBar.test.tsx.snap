// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 1`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      <button
        class="filterNavigatorBarItemContent profileFilterNavigator--tab-selector"
        type="button"
      >
        Full Range (⁨51ms⁩)
      </button>
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 2`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      <span
        class="filterNavigatorBarItemContent profileFilterNavigator--tab-selector"
      >
        Full Range (⁨51ms⁩)
      </span>
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      40ms
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 3`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      <span
        class="filterNavigatorBarItemContent profileFilterNavigator--tab-selector"
      >
        Full Range (⁨51ms⁩)
      </span>
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      40ms
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarUncommittedItem filterNavigatorBarLeafItem"
    title="100μs"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      100μs
    </span>
  </li>
</ol>
`;
